magicblock-committor-program/src/state/changeset.rs

```diff
@@ -0,0 +1,498 @@
+    /// The ephemeral slot at which those changes were requested
+    pub slot: Slot,
+    /// The accounts that should be undelegated after they were committed
+    pub accounts_to_undelegate: HashSet<Pubkey>,
+}
+
+/// The meta data of the changeset which can be used to capture information about
+/// the changeset before transferring ownership. Createing this metadata is
+/// a lot cheaper than copying the entire changeset which includes the accounts data.
+/// Thus it can be used to capture information to include with error responses.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct ChangesetMeta {
+    /// The accounts that should be updated
+    pub accounts: Vec<ChangedAccountMeta>,
+    /// The ephemeral slot at which those changes were requested
+    pub slot: Slot,
+    /// The accounts that should be undelegated after they were committed
+    pub accounts_to_undelegate: HashSet<Pubkey>,
+}
+
+impl ChangesetMeta {
+    /// Separates information per account including the following:
+    /// - account commit metadata
+    /// - slot at which commit was requested
+    /// - if the account should be undelegated after it was committed
+    pub fn into_account_infos(self) -> Vec<(ChangedAccountMeta, Slot, bool)> {
+        self.accounts
+            .into_iter()
+            .map(|account| {
+                let undelegate =
+                    self.accounts_to_undelegate.contains(&account.pubkey);
+                (account, self.slot, undelegate)
+            })
+            .collect()
+    }
+}
+
+impl From<&Changeset> for ChangesetMeta {
+    fn from(changeset: &Changeset) -> Self {
+        let accounts = changeset
+            .accounts
+            .iter()
+            .map(ChangedAccountMeta::from)
+            .collect();
+        Self {
+            accounts,
+            slot: changeset.slot,
+            accounts_to_undelegate: changeset.accounts_to_undelegate.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct ChangesetBundles {
+    /// The bundles, each of which needs to be committed atomically
+    pub bundles: Vec<ChangedBundle>,
+    /// The ephemeral slot at which those changes were requested
+    pub slot: Slot,
+    /// The accounts that should be undelegated after they were committed
+    pub accounts_to_undelegate: HashSet<Pubkey>,
+}
+
+impl Changeset {
+    /// Adds an account to the change set.
+    /// If it already exists, it will be replaced, thus the caller needs
+    /// to ensure that conflicting changes are added in the right order, i.e.
+    /// the last update needs to be added last.
+    ///
+    /// - **pubkey** public key of the account
+    /// - **account** account to add
+    ///
+    /// *returns* true if the account was already present and was replaced
+    pub fn add<T: Into<ChangedAccount>>(
+        &mut self,
+        pubkey: Pubkey,
+        account: T,
+    ) -> bool {
+        self.accounts.insert(pubkey, account.into()).is_some()
```

Comment:
> /// If it already exists, it will be replaced, thus the caller needs

I think that could be an issue, assume actions.
Imagine:
1. Bob makes a CommitWithActions with some accounts [acc1, acc2], and actions [a1, a2]. commit_id = 1
2. Bob makes a CommitWithActions with some accounts [acc2, acc3], and actions [a3, a4]. commit_id = 2
3. 
At this point we would replaced acc2 with commit_id = 1, to acc2 with commit_id = 2. As result for commit_id/bundle_id = 1 we would commit [acc1] with actions [a1, a2], when user expected both of them to be commited, and action could be connected to state of acc2.

Shouldn't we keep all accounts per commit_id, and not replace them? 
cc @GabrielePicco